File: src/backend/dotnet/TodoApi/Controllers/UsersController.cs
using System.Security.Claims;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Authorization;
using TodoApi.DTOs;
using TodoApi.Services;

namespace TodoApi.Controllers;

[ApiController]
[Route("api/users")]
[Produces("application/json")]
public class UsersController(IUserService service) : ControllerBase
{
    [HttpGet]
    [ProducesResponseType(typeof(PaginatedResponse<UserResponse>), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetAll([FromQuery] int page = 1, [FromQuery] int pageSize = 20, CancellationToken ct = default) =>
        Ok(await service.GetAllAsync(page, pageSize, ct));

    [HttpGet("{id:int}")]
    [ProducesResponseType(typeof(UserResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> GetById(int id, CancellationToken ct = default) =>
        await HandleNotFoundAsync(() => service.GetByIdAsync(id, ct));

    [HttpPost]
    [ProducesResponseType(typeof(UserResponse), StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    public async Task<IActionResult> Create([FromBody] CreateUserRequest request, CancellationToken ct = default)
    {
        try
        {
            var user = await service.CreateAsync(request, ct);
            return CreatedAtAction(nameof(GetById), new { id = user.Id }, user);
        }
        catch (UserConflictException ex)
        {
            return Conflict(new { error = ex.Message });
        }
    }

    [HttpPut("{id:int}")]
    public Task<IActionResult> Update(int id, [FromBody] UpdateUserRequest request, CancellationToken ct = default) =>
        HandleUserWriteAsync(() => service.UpdateAsync(id, request, ct));

    [HttpPatch("{id:int}/activate")]
    public Task<IActionResult> Activate(int id, CancellationToken ct = default) =>
        HandleUserWriteAsync(() => service.SetActiveAsync(id, true, ct));

    [HttpPatch("{id:int}/deactivate")]
    public Task<IActionResult> Deactivate(int id, CancellationToken ct = default) =>
        HandleUserWriteAsync(() => service.SetActiveAsync(id, false, ct));

    [HttpPost("signup")]
    [ProducesResponseType(typeof(UserResponse), StatusCodes.Status201Created)]
    public async Task<IActionResult> SignUp([FromBody] SignUpRequest request, CancellationToken ct = default)
    {
        try
        {
            var user = await service.SignUpAsync(request, ct);
            return CreatedAtAction(nameof(GetById), new { id = user.Id }, user);
        }
        catch (UserConflictException ex)
        {
            return Conflict(new { error = ex.Message });
        }
    }

    [HttpGet("profile")]
    [Authorize]
    public async Task<IActionResult> GetProfile(CancellationToken ct = default)
    {
        if (!TryGetCurrentUserId(out var userId)) return Unauthorized();
        return await HandleNotFoundAsync(() => service.GetProfileAsync(userId, ct));
    }

    [HttpPut("profile")]
    [Authorize]
    public async Task<IActionResult> UpdateProfile([FromBody] UpdateProfileRequest request, CancellationToken ct = default)
    {
        if (!TryGetCurrentUserId(out var userId)) return Unauthorized();
        return await HandleUserWriteAsync(() => service.UpdateProfileAsync(userId, request, ct));
    }

    [HttpPost("password/change")]
    [Authorize]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    public async Task<IActionResult> ChangePassword([FromBody] ChangePasswordRequest request, CancellationToken ct = default)
    {
        if (!TryGetCurrentUserId(out var userId)) return Unauthorized();
        try
        {
            await service.ChangePasswordAsync(userId, request, ct);
            return NoContent();
        }
        catch (KeyNotFoundException ex)
        {
            return NotFound(new { error = ex.Message });
        }
        catch (InvalidPasswordException ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    [HttpPost("password/reset")]
    [ProducesResponseType(StatusCodes.Status202Accepted)]
    public async Task<IActionResult> RequestPasswordReset([FromBody] ResetPasswordRequest request, CancellationToken ct = default)
    {
        await service.RequestPasswordResetAsync(request, ct);
        return Accepted(new { message = "If the account exists, a password reset email has been queued." });
    }

    [HttpPost("password/confirm")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    public async Task<IActionResult> ConfirmPasswordReset([FromBody] ConfirmPasswordResetRequest request, CancellationToken ct = default)
    {
        try
        {
            await service.ConfirmPasswordResetAsync(request, ct);
            return NoContent();
        }
        catch (InvalidPasswordResetTokenException ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    private bool TryGetCurrentUserId(out int userId)
    {
        var principal = HttpContext?.User;
        var value = principal?.FindFirstValue(ClaimTypes.NameIdentifier) ?? principal?.FindFirstValue("sub");
        return int.TryParse(value, out userId) && userId > 0;
    }

    private static async Task<IActionResult> HandleNotFoundAsync(Func<Task<UserResponse>> action)
    {
        try { return new OkObjectResult(await action()); }
        catch (KeyNotFoundException ex) { return new NotFoundObjectResult(new { error = ex.Message }); }
    }

    private static async Task<IActionResult> HandleUserWriteAsync(Func<Task<UserResponse>> action)
    {
        try { return new OkObjectResult(await action()); }
        catch (KeyNotFoundException ex) { return new NotFoundObjectResult(new { error = ex.Message }); }
        catch (UserConflictException ex) { return new ConflictObjectResult(new { error = ex.Message }); }
    }
}

File: src/backend/nodejs/src/api/users/users.controller.ts
import { Body, Controller, Get, HttpCode, HttpStatus, Param, ParseIntPipe, Patch, Post, Put, Query, Req, UseGuards } from '@nestjs/common';
import { ApiAcceptedResponse, ApiBearerAuth, ApiCreatedResponse, ApiNoContentResponse, ApiOkResponse, ApiQuery, ApiTags } from '@nestjs/swagger';
import { PaginatedResponseDto } from '../../shared/common/dto/paginated-response.dto';
import { CreateUserDto, SignUpDto } from './dto/create-user.dto';
import { ChangePasswordDto, ConfirmPasswordResetDto, ResetPasswordDto } from './dto/password.dto';
import { UpdateProfileDto, UpdateUserDto } from './dto/update-user.dto';
import { UserResponseDto } from './dto/user-response.dto';
import { AuthenticatedRequest, UserAuthGuard } from './users.security';
import { UsersService } from './users.service';

@ApiTags('Users')
@Controller('api/users')
export class UsersController {
  constructor(private readonly service: UsersService) {}

  @Post('signup')
  @HttpCode(HttpStatus.CREATED)
  @ApiCreatedResponse({ type: UserResponseDto })
  signup(@Body() dto: SignUpDto): Promise<UserResponseDto> {
    return this.service.signup(dto);
  }

  @Post('password/reset')
  @HttpCode(HttpStatus.ACCEPTED)
  @ApiAcceptedResponse({ description: 'Password reset request accepted' })
  async requestPasswordReset(@Body() dto: ResetPasswordDto): Promise<{ message: string }> {
    await this.service.requestPasswordReset(dto);
    return { message: 'If the account exists, a password reset email has been queued.' };
  }

  @Post('password/confirm')
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiNoContentResponse()
  confirmPasswordReset(@Body() dto: ConfirmPasswordResetDto): Promise<void> {
    return this.service.confirmPasswordReset(dto);
  }

  @Post('password/change')
  @UseGuards(UserAuthGuard)
  @ApiBearerAuth()
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiNoContentResponse()
  changePassword(@Req() request: AuthenticatedRequest, @Body() dto: ChangePasswordDto): Promise<void> {
    return this.service.changePassword(request.user.userId, dto);
  }

  @Get('profile')
  @UseGuards(UserAuthGuard)
  @ApiBearerAuth()
  @ApiOkResponse({ type: UserResponseDto })
  getProfile(@Req() request: AuthenticatedRequest): Promise<UserResponseDto> {
    return this.service.getProfile(request.user.userId);
  }

  @Put('profile')
  @UseGuards(UserAuthGuard)
  @ApiBearerAuth()
  @ApiOkResponse({ type: UserResponseDto })
  updateProfile(@Req() request: AuthenticatedRequest, @Body() dto: UpdateProfileDto): Promise<UserResponseDto> {
    return this.service.updateProfile(request.user.userId, dto);
  }

  @Get()
  @ApiQuery({ name: 'page', required: false, type: Number })
  @ApiQuery({ name: 'pageSize', required: false, type: Number })
  @ApiOkResponse({ description: 'Paginated list of users' })
  getAll(
    @Query('page', new ParseIntPipe({ optional: true })) page = 1,
    @Query('pageSize', new ParseIntPipe({ optional: true })) pageSize = 20,
  ): Promise<PaginatedResponseDto<UserResponseDto>> {
    return this.service.getAll(page, pageSize);
  }

  @Post()
  @HttpCode(HttpStatus.CREATED)
  @ApiCreatedResponse({ type: UserResponseDto })
  create(@Body() dto: CreateUserDto): Promise<UserResponseDto> {
    return this.service.create(dto);
  }

  @Get(':id')
  @ApiOkResponse({ type: UserResponseDto })
  getById(@Param('id', ParseIntPipe) id: number): Promise<UserResponseDto> {
    return this.service.getById(id);
  }

  @Put(':id')
  @ApiOkResponse({ type: UserResponseDto })
  update(@Param('id', ParseIntPipe) id: number, @Body() dto: UpdateUserDto): Promise<UserResponseDto> {
    return this.service.update(id, dto);
  }

  @Patch(':id/activate')
  @ApiOkResponse({ type: UserResponseDto })
  activate(@Param('id', ParseIntPipe) id: number): Promise<UserResponseDto> {
    return this.service.setActive(id, true);
  }

  @Patch(':id/deactivate')
  @ApiOkResponse({ type: UserResponseDto })
  deactivate(@Param('id', ParseIntPipe) id: number): Promise<UserResponseDto> {
    return this.service.setActive(id, false);
  }
}
